src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/SpellCheck/SpellCheckService.cs

```diff
@@ -0,0 +1,137 @@
+                // parameters as strings, or they're Html attributes which are not necessarily expected to be real words.
+                if (node.Parent is MarkupTagHelperAttributeSyntax or
+                    MarkupAttributeBlockSyntax or
+                    MarkupMinimizedAttributeBlockSyntax or
+                    MarkupTagHelperDirectiveAttributeSyntax or
+                    MarkupMinimizedTagHelperAttributeSyntax or
+                    MarkupMinimizedTagHelperDirectiveAttributeSyntax or
+                    MarkupMiscAttributeContentSyntax)
+                {
+                    continue;
+                }
+
+                // Text literals appear everywhere in Razor to hold newlines and indentation, so its worth saving the tokens
+                if (textLiteralSyntax.ContainsOnlyWhitespace())
+                {
+                    continue;
+                }
+
+                if (textLiteralSyntax.Span.Length == 0)
+                {
+                    continue;
+                }
+
+                ranges.Add(new((int)VSInternalSpellCheckableRangeKind.String, textLiteralSyntax.SpanStart, textLiteralSyntax.Span.Length));
+            }
+        }
+    }
+
+    private void AddCSharpSpellCheckRanges(ref PooledArrayBuilder<SpellCheckRange> ranges, ImmutableArray<SpellCheckRange> csharpRanges, RazorCodeDocument codeDocument)
+    {
+        var csharpDocument = codeDocument.GetCSharpDocument();
+
+        foreach (var report in csharpRanges)
+        {
+            var absoluteCSharpStartIndex = report.AbsoluteStartIndex;
+            var length = report.Length;
+
+            // We need to map the start index to produce results, and we validate that we can map the end index so we don't have
+            // squiggles that go from C# into Razor/Html.
+            if (_documentMappingService.TryMapToHostDocumentPosition(csharpDocument, absoluteCSharpStartIndex, out var _1, out var hostDocumentIndex) &&
+                _documentMappingService.TryMapToHostDocumentPosition(csharpDocument, absoluteCSharpStartIndex + length, out var _2, out var _3))
+            {
+                ranges.Add(new(report.Kind, hostDocumentIndex, length));
+            }
+        }
+    }
+
+    private static int[] ConvertSpellCheckRangesToIntTriples(ImmutableArray<SpellCheckRange> ranges)
+    {
+        // Important to sort first as we're calculating relative indexes
+        ranges = ranges.OrderAsArray(CompareSpellCheckRanges);
+
+        using var data = new PooledArrayBuilder<int>(ranges.Length * 3);
+
+        var lastAbsoluteEndIndex = 0;
+        foreach (var range in ranges)
+        {
+            if (range.Length == 0)
+            {
+                continue;
+            }
+
+            data.Add(range.Kind);
+            data.Add(range.AbsoluteStartIndex - lastAbsoluteEndIndex);
+            data.Add(range.Length);
+
+            lastAbsoluteEndIndex = range.AbsoluteStartIndex + range.Length;
+        }
```

Comment:
Consider using deconstructing to assign to locals:

```suggestion
        {
            var (kind, absoluteStartIndex, length) = range;
            if (length == 0)
            {
                continue;
            }

            data.Add(kind);
            data.Add(absoluteStartIndex - lastAbsoluteEndIndex);
            data.Add(length);

            lastAbsoluteEndIndex = absoluteStartIndex + length;
        }
```

Reply:
How strongly do you fell about it? As mentioned above, this makes me nervous.

Reply:
I'm actually not sure how to respond. Your choice of words in this comment and the other leaves me wondering if you're a bit annoyed at my suggestion. If positional record (and, by extension, tuple) deconstruction is considered harmful, perhaps using positional records at all should be reconsidered?

If you're not comfortable with this, definitely don't do it. I won't push the issue and won't leave such feedback in the future. To me, it's no different than calling a method that takes three `int` parameters. Yes, they can get out of order if a programmer is not careful, but is the answer to always use named arguments at every call site to be resilient against reordering?

Reply:
Definitely not annoyed at the suggestion, and I'm much too pragmatic to let a specific scenario make me want to rule out a language feature altogether.

For me there are subtle but important distinctions in this case, versus others. I will try to explain my thinking, but if the arguments end up sounding arbitrary, I won't be surprised :)

1. No IDE help. `var (a, b, c) = foo;` provides nothing I can F12 on or hover over to see which Deconstruct method is being called, and what the parameters are. Ditto even a type or method name to search for in GitHub.
2. No context clues: because these are all the same type, none of the other code here provides hints or reassurance that they are being deconstructed in the right order. If one of them was a string and one a DateTime for example, I'd be more happy to trust that the order is correct because other code in the method would fail to compile.
3. Not a common scenario. If this were deconstructing a common type, like `KeyValuePair<int, int>` or `Position`/`LinePosition` then it would be reasonable to assume the reader is able to recognize the shape of the deconstruction regardless of the above two points. OTOH I've already forgotten what's in a SpellCheckRange.

Reply:
> and won't leave such feedback in the future

That's entirely up to you, but after typing all that out (on my phone so perhaps with less forethought than is warranted 😛) I do wonder if my opinions on this will change in time. I can see a little of the same arguments that were made against `var` when it was new, about a change in method signature/type definition not causing a compile time error, at every call site.

Maybe I just need exposure therapy and I'll be convinced that the Deconstruct method can't hurt me :)

Reply:
> In the code here, this problem of data shape and having no help from the IDE applies even more strongly to the order of the "triple" data in the int array

Yes, 100%, but thats in the LSP spec so we just have to deal with it.

> You could have used slicing and a list pattern to make things a little clearer to my eye

I do like this, but my brain is slow to adapt and I do not yet reach for spans and slicing etc. when trying to solve problems.